File: coreAlgorithms/binarySearch.py
'''
Author: chenbei
Date: 2022-03-15 09:04:38
LastEditTime: 2022-03-15 10:00:10
Description: binary search
FilePath: \myLeetCode\coreAlgorithms\binarySearch.py
A boy without dreams
'''
from typing import Any, List
class binarySearch:
    def __init__(self,nums:List[int or float],target:int or float) -> Any:
        self.nums = nums
        self.nums.sort()
        self.target = target
    def search(self) -> int:
        left,right = 0,len(self.nums)-1
        while left<=right:
            mid = (left+right) // 2
            if self.nums[mid] == self.target:
                return mid
            if self.nums[mid] < self.target: # target∈[mid+1,right]
                left = mid + 1
            if self.nums[mid] > self.target: # target∈[left,mid-1]
                right = mid - 1
        return -1
    def sorted(self,reverse=False)->List[int or float]:
        if reverse:
            self.nums.sort(reverse=True)
        return self.nums
if __name__ == '__main__':
    solution1 = binarySearch([1.5,2,3.1,4.5,6.7,9.4,-12,-5.2],9.4)
    ans = solution1.sorted()
    print(ans)
    print("9.4's idx in sorted nums is ",solution1.search())
    solution2 = binarySearch([1.5,2,3.1,4.5,6.7,9.4,-12,-5.2],9.3)
    print("9.3's idx in sorted is ",solution2.search())

File: STLTestCode/algorithms/source/binary_search.cpp
/*** 
 * @Author: chenbei
 * @Date: 2022-04-22 08:22:08
 * @LastEditTime: 2022-04-22 08:37:23
 * @Description: test binary_search.cpp
 * @FilePath: \myLeetCode\STLTestCode\algorithms\source\binary_search.cpp
 * @Signature: A boy without dreams
 */
/*
g++ -o ..\output\binary_search binary_search.cpp
..\output\binary_search.exe
*/
#include <algorithm>
#include <iostream>
#include <vector>
#include <iterator>
using namespace std;
int main()
{
    // 找到这个值返回true,否则返回false
    vector<int> src = {1,2,3,4,5,6,7,8,9,10};
    int value = 5;
    auto it = binary_search(src.begin(), src.end(), value);
    if(it)
    {
        cout << "find " << value << endl; // find 5
    }
    else
    {
        cout << "not find " << value << endl;
    }
    // 指定函数对象也可以,但必须和用于对被查找序列进行排序的比较操作有相同的效果
    // 现在容器是升序的,但是传入的比较函数对象是降序的
    // 所以返回的是false,而不是true
    auto it2 = binary_search(src.begin(), src.end(), value,greater<int>());
    if(it2)
    {
        cout << "find " << value << endl;
    }
    else
    {
        cout << "not find " << value << endl; // not find 5
    }
    return 0;
}